server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -591,86 +725,67 @@ public void visit(int docID, byte[] packedValue) throws IOException {
+                if (!collector.withinUpperBound(packedValue)) {
                     collector.finalizePreviousRange();
-                    if (collector.iterateRangeEnd(packedValue, this::compareByteValue)) {
+                    if (collector.iterateRangeEnd(packedValue)) {
                         throw new CollectionTerminatedException();
                     }
                 }
 
-                if (pointCompare(collector.activeRangeAsByteArray[0], collector.activeRangeAsByteArray[1], packedValue)) {
+                if (collector.withinRange(packedValue)) {
                     collect.run();
                 }
             }
 
-            private boolean pointCompare(byte[] lower, byte[] upper, byte[] packedValue) {
-                if (compareByteValue(packedValue, lower) < 0) {
-                    return false;
-                }
-                return compareByteValue(packedValue, upper) <= 0;
-            }
-
-            private int compareByteValue(byte[] value1, byte[] value2) {
-                return comparator.compare(value1, 0, value2, 0);
-            }
-
             @Override
             public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
-                byte[] rangeMin = collector.activeRangeAsByteArray[0];
-                byte[] rangeMax = collector.activeRangeAsByteArray[1];
-
-                if (compareByteValue(rangeMax, minPackedValue) < 0) {
+                // try to find the first range that may collect values from this cell
+                if (!collector.withinUpperBound(minPackedValue)) {
                     collector.finalizePreviousRange();
-                    if (collector.iterateRangeEnd(minPackedValue, this::compareByteValue)) {
+                    if (collector.iterateRangeEnd(minPackedValue)) {
                         throw new CollectionTerminatedException();
                     }
-                    // compare the next range with this node's min max again
-                    // new rangeMin = previous rangeMax + 1 <= min
-                    rangeMax = collector.activeRangeAsByteArray[1];
                 }
-
-                if (compareByteValue(rangeMin, minPackedValue) > 0 || compareByteValue(rangeMax, maxPackedValue) < 0) {
-                    return PointValues.Relation.CELL_CROSSES_QUERY;
-                } else {
+                // after the loop, min < upper
+                // cell could be outside [min max] lower
+                if (!collector.withinLowerBound(maxPackedValue)) {
+                    return PointValues.Relation.CELL_OUTSIDE_QUERY;
+                }
+                if (collector.withinRange(minPackedValue) && collector.withinRange(maxPackedValue)) {
                     return PointValues.Relation.CELL_INSIDE_QUERY;
                 }
+                return PointValues.Relation.CELL_CROSSES_QUERY;
             }
         };
     }
 
     private static class RangeCollectorForPointTree {
```

Comment:
While some of these classes can/should also be moved outside `FastFilterRewriteHelper`, I am okay for now since they are generic/aggregation-agnostic. Let us refactor any aggregation type specific logic from this class

Reply:
Can cover in the refactor PR.